crates/ruma-common/src/room_version_rules.rs

```diff
@@ -0,0 +1,442 @@
+pub enum RoomVersionDisposition {
+    /// A room version that has a stable specification.
+    Stable,
+
+    /// A room version that is not yet fully specified.
+    Unstable,
+}
+
+/// The format of [event IDs] for a room version.
+///
+/// [event IDs]: https://spec.matrix.org/latest/appendices/#event-ids
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub enum EventIdFormatVersion {
+    /// `$id:server` format ([spec]), introduced in room version 1.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v1/#event-ids
+    V1,
+
+    /// `$hash` format using standard unpadded base64 ([spec]), introduced in room version 3.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v3/#event-ids
+    V2,
+
+    /// `$hash` format using URL-safe unpadded base64 ([spec]), introduced in room version 4.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v4/#event-ids
+    V3,
+}
+
+/// The version of [state resolution] for a room version.
+///
+/// [state resolution]: https://spec.matrix.org/latest/server-server-api/#room-state-resolution
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub enum StateResolutionVersion {
+    /// First version of the state resolution algorithm ([spec]), introduced in room version 1.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v1/#state-resolution
+    V1,
+
+    /// Second version of the state resolution algorithm ([spec]), introduced in room version 2.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v2/#state-resolution
+    V2,
+}
+
+/// The tweaks in the [authorization rules] for a room version.
+///
+/// This type can be constructed from one of its constants (like [`AuthorizationRules::V1`]), or by
+/// constructing a [`RoomVersionRules`] first and using the `authorization` field.
+///
+/// [authorization rules]: https://spec.matrix.org/latest/server-server-api/#authorization-rules
+#[derive(Debug, Clone)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct AuthorizationRules {
+    /// Whether to apply special authorization rules for `m.room.redaction` events ([spec]),
+    /// disabled since room version 3.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v3/#handling-redactions
+    pub special_case_room_redaction: bool,
+
+    /// Whether to apply special authorization rules for `m.room.aliases` events ([spec]), disabled
+    /// since room version 6.
+    ///
+    /// [spec]: https://spec.matrix.org/latest/rooms/v6/#authorization-rules
+    pub special_case_room_aliases: bool,
+
+    /// Whether to strictly enforce [canonical JSON] ([spec]), introduced in room version 6.
+    ///
+    /// Canonical JSON does not allow:
+    ///
+    /// * Integers outside the range of [-2<sup>53</sup> + 1, 2<sup>53</sup> - 1]
+    /// * Floats
+    /// * `NaN`, `Infinity`, `-Infinity`
```

Comment:
Exponents fall under floats, right?
```suggestion
    /// * Floats
    /// * `NaN`, `Infinity`, `-Infinity`, `-0`
```

Reply:
I am not sure that they do.

However to me `NaN`, `Infinity` and `-Infinity` do. And besides they are clearly outside of the range in the first item…

Let me fix that.

Reply:
Alright, I took the text from the spec, it's simpler.